cond diff (alt fix)

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92644

Fixes https://github.com/pytorch/pytorch/issues/92561

When different branches of a cond reference different functions in outer scope, a spurious diff of side effects may be created because (1) we create new cells for such closure variables (2) we prune these cells per branch based on what's accessed and what's not.

Alternative fix here: https://github.com/pytorch/pytorch/pull/92565

Differential Revision: [D42617685](https://our.internmc.facebook.com/intern/diff/D42617685/)

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

avikchaudhuri committed:
cond diff (alt fix)

Fixes https://github.com/pytorch/pytorch/issues/92561

When different branches of a cond reference different functions in outer scope, a spurious diff of side effects may be created because (1) we create new cells for such closure variables (2) we prune these cells per branch based on what's accessed and what's not.

Alternative fix here: https://github.com/pytorch/pytorch/pull/92565

Differential Revision: [D42617685](https://our.internmc.facebook.com/intern/diff/D42617685/)

[ghstack-poisoned]

ezyang commented on torch/_dynamo/side_effects.py:
```diff
@@ -246,8 +246,9 @@ def track_global_existing(self, source: Source, item: Any):
-        live_new_objects = set()
+    def prune_dead_object_new(self, tx, live_new_objects=None):
+        if live_new_objects is None:
+            live_new_objects = set()
```
Careful, do you want to mutate the input?

avikchaudhuri replied:
Why not, isn't this a harmless pattern for setting the default of what is now an optional param `live_new_objects`? You either pass a set explicitly, or you omit it and it gets initialized as before.

avikchaudhuri replied:
Oh I see where this question is coming from now, and I think it's related to the @jansel question below as well. I *do* want to mutate this set. This is just an initial set, it is expected to grow. The intent is that *at least* the objects is this set will not be pruned, although fewer objects might end up being pruned because of additional marking in this function as before.

ezyang reviewed (approved):
I like this a lot better. It's not entirely clear to me if it is still correct, but at least if we're wrong it will fail with an explicit error.

jansel commented on torch/_dynamo/variables/torch.py:
```diff
@@ -761,7 +761,15 @@ def speculate_branch(branch):
+                    # These "side effects" should not be pruned, o/w it causes a diff between branches.
+                    # We shouldn't care when branches call different sets of functions in outer scope.
+                    live_new_objects=set(
+                        checkpoint.output.side_effects.store_attr_mutations.keys()
```
Do you have an example where we hit this?  I would not expect the example above to generate STORE_ATTR?

If we had a conditional STORE_ATTR I worry things would just be wrong.  For example:
```
class Obj:
  def forward(self, x):
     return cond(x>0, self.if_true, self.if_false, [x])

  def if_true(self, x):
     self.foo = 2  # STORE_ATTR
     return x

  def if_false(self, x):
     self.foo = 1  # STORE_ATTR
     return x
```

In this case dynamo will normally move the STORE_ATTR to the end of the graph, but in this case it will be moved to the end of the subgraph then perhaps dropped?

This seems super hard to deal with, but also really uncommon.  I'd almost just add an assert that it doesn't happen and move on.

avikchaudhuri replied:
The added test in `test_export` is the example I'm using (representative repro of part of an internal model, see task attached to internal diff).

It has functions defined inside a method, and those end up in `store_attr_mutations`. I can try to dig in how. Maybe the nested functions get stored as attributes in the parent module?

Regardless, I didn't follow your concern here, can you please elaborate? (Sorry I don't have the full mental model of what's supposed to happen here yet.) Here's my reasoning for why this diff is not changing behavior here, please correct me if I'm wrong.
1. Since `checkpoint` refers to the state outside `if_true` / `if_false` (before we enter `speculate_branch`), I'd guess `self.foo` in your example would not be passed as part of the initial set of live new objects here. 
2. Note that inside `prune_dead_object_new` this set will grow as usual (the only difference is, before it started from an empty set, now we might have some initial set). So if `self.foo` ended up being marked live before this diff, it would after this diff as well, and it wouldn't be dropped.

In other words, I'm trying to *avoid* some objects being dropped, not dropping any object that wasn't dropped before.

avikchaudhuri replied:
See also reply to @ezyang above. If my guess is right about the source of confusion, I'll add a comment to the pruning function / rename the optional param that initializes the live set.

jansel replied:
I believe we want to:
1) Checkpoint side_effects at the start of `speculate_branch`
2) Possibly assert something about side effects (no side effects have happened, closures haven't been mutated, etc)
3) Restore the checkpoint to get side_effects to the initial state (at the end of the speculate_branch)

If `speculate_branch` mutates side_effects, something is going wrong.  It would mean state could be leaking out of speculatively executed code.

Specifically I want this code to throw an error:
```
  z = 2

  def true_fn(y):
      nonlocal z
      z = 99
      return y + 2
  
  def false_fn(y):
      return y - z
```


I worry right now running `speculate_branch(true_fn)` will mutate the global state in `side_effects`

avikchaudhuri replied:
There's already saving / restoring of checkpoints during speculation (side effects are part of the output graph state, which is the checkpoint)! And indeed the code above errors. 

The error is the diff of the states at the end of the two branches. Specifically, symbolic_locals['z'] points to 99 and 2, respectively, which causes the error.

The only difference between current behavior and what you state above is that there's no assertion against state mutation, rather only against a diff in state mutation.

I'm not opposed to us changing the implementation of closures / state for export, but *I repeat that it's orthogonal to this PR*. 

The bug I'm specifically fixing here is that *pruning* in an inner scope obliterates new cells it doesn't even own (because they belong to the outer scope), which in turn causes a spurious diff. OTOH what you're concerned with above has no relation to *pruning*. Is that right?

ezyang replied:
I feel like part of the confusion is a disagreement about what the data structures look like in the example. In particular, I don't think jansel has tried running the example yet. Maybe it would be helpful to produce some debug prints of the data structure state here.

jansel replied:
Yeah, some more concrete examples would be helpful.  I would not expect this PR to need to change side effects.

avikchaudhuri committed:
Update on "cond diff (alt fix)"


Fixes https://github.com/pytorch/pytorch/issues/92561

When different branches of a cond reference different functions in outer scope, a spurious diff of side effects may be created because (1) we create new cells for such closure variables (2) we prune these cells per branch based on what's accessed and what's not.

Alternative fix here: https://github.com/pytorch/pytorch/pull/92565

Differential Revision: [D42617685](https://our.internmc.facebook.com/intern/diff/D42617685/)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri committed:
Update on "cond diff (alt fix)"


Fixes https://github.com/pytorch/pytorch/issues/92561

When different branches of a cond reference different functions in outer scope, a spurious diff of side effects may be created because (1) we create new cells for such closure variables (2) we prune these cells per branch based on what's accessed and what's not.

Alternative fix here: https://github.com/pytorch/pytorch/pull/92565

Differential Revision: [D42617685](https://our.internmc.facebook.com/intern/diff/D42617685/)

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot rebase

jansel reviewed (changes requested):
RQ to take this off my review queue.  I believe we have already discussed this.  Feel free to re-request if you want me to take another look.
